### PR TITLE
roachprod: preserve error object in Get

### LINKED ISSUE #132434: roachtest: jepsen/bank-multitable/parts-start-kill-2 failed
roachtest.jepsen/bank-multitable/parts-start-kill-2 [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/17270349?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/17270349?buildTab=artifacts#/jepsen/bank-multitable/parts-start-kill-2) on release-24.2 @ [5405e9098bbc9e12ccc4387664eaedfc2d32e9d9](https://github.com/cockroachdb/cockroach/commits/5405e9098bbc9e12ccc4387664eaedfc2d32e9d9):

```
(jepsen.go:177).initJepsen: cluster.Get: get .ssh/id_rsa.pub failed
test artifacts and logs in: /artifacts/jepsen/bank-multitable/parts-start-kill-2/cpu_arch=arm64/run_1
```

Parameters:
 - <code>ROACHTEST_arch=arm64</code>
 - <code>ROACHTEST_cloud=gce</code>
 - <code>ROACHTEST_coverageBuild=false</code>
 - <code>ROACHTEST_cpu=4</code>
 - <code>ROACHTEST_encrypted=false</code>
 - <code>ROACHTEST_fs=ext4</code>
 - <code>ROACHTEST_localSSD=false</code>
 - <code>ROACHTEST_runtimeAssertionsBuild=false</code>
 - <code>ROACHTEST_ssd=0</code>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-17270349/jepsen-bank-multitable-parts-start-kill-2/1728647310322/1728650288801)

</p>
</details>
/cc @cockroachdb/test-eng
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*jepsen/bank-multitable/parts-start-kill-2.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

Jira issue: CRDB-43063

### PR DESCRIPTION
**Pull request by DarrylWong**
We need to keep the error object so our ssh flake detection works.

Fixes: #132434
Release note: none
Epic: none

### CODE REVIEW

**Comment by DarrylWong** [comment]
Before this change if the `scp` in `Get` timed out we got:

```
2024/10/11 18:58:02 test_impl.go:422: test failure #1: full stack retained in failure_1.log: (jepsen.go:178).initJepsen: cluster.Get: get .ssh/id_rsa.pub failed
```

With this change, we now get:

```
2024/10/11 18:50:29 test_impl.go:422: test failure #1: full stack retained in failure_1.log: (jepsen.go:178).initJepsen: cluster.Get: get .ssh/id_rsa.pub failed: TRANSIENT_ERROR(ssh_problem): ~ scp -r -C -o StrictHostKeyChecking=no -o ConnectTimeout=10 -R -A -i /home/darryl_cockroachlabs_com/.ssh/id_ed25519 ubuntu@34.75.78.35:.ssh/id_rsa.pub /tmp/jepsen1466720316/id_rsa.pub: context deadline exceeded
 ```

File: pkg/roachprod/install/cluster_synced.go [hunk 1]
```diff
@@ -2597,24 +2597,33 @@ func (c *SyncedCluster) Get(
 		close(results)
 	}()
 
+	var errOnce sync.Once
+	var finalErr error
+	setErr := func(e error) {
+		if e != nil {
```
Comment (herkolategan) [inline pkg/roachprod/install/cluster_synced.go:2603]: Nit: There's already a `r.err != nil` check, so this one might not be necessary.

File: pkg/roachprod/install/cluster_synced.go [hunk 2]
```diff
@@ -2597,24 +2597,33 @@ func (c *SyncedCluster) Get(
...
+			errOnce.Do(func() {
```
Comment (herkolategan) [inline pkg/roachprod/install/cluster_synced.go:2604]: Just for my own understanding, is there a particular reason we capture only the first error, does it matter if this becomes the last error?

File: pkg/roachprod/install/cluster_synced.go [hunk 3]
```diff
@@ -2597,24 +2597,33 @@ func (c *SyncedCluster) Get(
 		close(results)
 	}()
 
+	var errOnce sync.Once
```
Comment (herkolategan) [inline pkg/roachprod/install/cluster_synced.go:2600]: I don't think this is necessarily used in a concurrent access context (just a regular for loop on a channel, unless I missed something), so you might be able to get away with just checking if the value has been set already, if that is the intention (see questionary comment below)

**Review by herkolategan — APPROVED** [review]
Small nit & questions. Thanks for fixing and testing that it works as intended!

File: pkg/roachprod/install/cluster_synced.go (hunk 1 above)
Comment (DarrylWong) [reply pkg/roachprod/install/cluster_synced.go]: Removed.

File: pkg/roachprod/install/cluster_synced.go (hunk 2 above)
Comment (DarrylWong) [reply pkg/roachprod/install/cluster_synced.go]: Yeah admittedly I'm not too sure either, I was just copying the behavior found in `Put()`. The commit for that doesn't give much insight either so I just played it safe and kept it.

>does it matter if this becomes the last error?

I'm guessing no and it just chooses one of the errors so as to not spam the user with an error per node. And maybe choosing the first one is more intentional than letting it repeatedly overwrite an err object 🤷

File: pkg/roachprod/install/cluster_synced.go (hunk 3 above)
Comment (DarrylWong) [reply pkg/roachprod/install/cluster_synced.go]: Yeah that sounds right to me, removing the `sync`.

File: pkg/roachprod/install/cluster_synced.go (hunk 2 above)
Comment (herkolategan) [reply pkg/roachprod/install/cluster_synced.go]: Ah makes sense, I see the logic came from somewhere else. Thanks for updating!

**Comment by DarrylWong** [comment]
TFTR!

bors r=herkolategan

**Pull request merged**